Add Dark Mode buttons fix for non-black dark themes;

Leonard-The-Wise opened the pull request:
Hello @donho,

I've added a patch to fix the dialog buttons, so they won't draw black everytime on custom colored Dark Themes.

Original look:

![image](https://user-images.githubusercontent.com/99574879/165865540-8629038f-7033-46e9-b557-540cb4be460b.png)

Patched:

![image](https://user-images.githubusercontent.com/99574879/166598149-02b6667e-04c4-496e-84df-a4aab08d0655.png)

I also modified the current (new) spin control to draw more seemingly to the original control - also now supports custom colors; also, now it will not re-subclass the control everytime it changes sizes. It will do it only once (at initialization).

Original unthemed version:

![image](https://user-images.githubusercontent.com/99574879/165865724-37a2df79-2849-4a13-90ff-95d4bb992d59.png)

The current version in repository:

![image](https://user-images.githubusercontent.com/99574879/165865820-88f098e5-1ee6-4b5f-aca1-d4718b1de8c4.png)

My version:

![image](https://user-images.githubusercontent.com/99574879/165866475-f508f2aa-e04d-4d25-a7ca-0c189218324e.png)


~~Lastly, I reverted the change to add new member to `struct Colors`, because that color is just derived from the existing `colors.edge`; no new customizable option exist. So It would be good to avoid breaking compatibility for `NPPM_GETDARKMODECOLORS` message unless the new color can be customized by the user in preferences dialog, instead of being a derived color. If you care to know how I achieved the same results without defining a custom color, I just added this function:~~ (not anymore. I saw the pull request to add the customized color).

Cheers.

sXNA commented:
Hi my first time writing here.

Great work and looks good, but can you change default background color for button to look more like button in Windows (before this commit), this commit makes it too dark. It looks like it's disabled.

I am using both Windows 10 and 11, and I don't want round corners on Windows 10 only on 11, could you fix this. 
![image](https://user-images.githubusercontent.com/30736899/166627740-afb1ac9b-57b9-43d6-aec0-c838562bad45.png)
![image](https://user-images.githubusercontent.com/30736899/166627768-7e0e5a17-68fd-44be-a845-18633bdf74d0.png)

Also this is why I prefer current tab overflow looks.

Current looks on Windows 10.
![image](https://user-images.githubusercontent.com/30736899/166629193-07b34ef7-623c-43c6-8fb9-551b29f3f05a.png)

Can you add custom color for highlighted border, I want to change it,  

btw
I am learning c++, and have question.
Doesn't check for `if (hwndUpdown == nullptr);` avoid resubclassing.

Leonard-The-Wise commented:
Yeah, the button background was actually a bug, thanks for pointing.
I also made it square and modified the arrow - for windows 10 of course. I'm afraid Windows 11 have to stay the same for consistency though.

![image](https://user-images.githubusercontent.com/99574879/166658796-99442073-7319-4446-a401-88cb17506a61.png)

![image](https://user-images.githubusercontent.com/99574879/166658855-3bb07c01-4563-4f2a-a660-2363d65fa9d4.png)


Also, for your questions:

- Can you add custom color for highlighted border, I want to change it,
Only if Don wants to add a customized color for that. You know, it would imply on an extra option here and I think colors are already harmonic the way they are... Changing the border edge just for hot buttons could break that harmony. Don't know, Don's call on that.

![image](https://user-images.githubusercontent.com/99574879/166659060-5b679815-3e33-460d-80c4-4dcd92139764.png)


Also

- Doesn't check for `if (hwndUpdown == nullptr);` avoid resubclassing.

Don't work for tabs... they create and destroy their own spin controls at run time. Not sure, but I think `hwndUpdown` would keep changing and changing also. Rather subclass once and forget about it (and I don't think subclassing itself is a bad thing - if Microsoft did it right, it can be very cheap - processing time wise).

Leonard-The-Wise commented:
Undone the change to not add customized color to hot edge to conform to [this pull request](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/11641)

donho commented on PowerEditor/src/NppDarkMode.cpp:
```diff
@@ -1184,9 +1520,9 @@ namespace NppDarkMode
 	}
 
-	constexpr UINT_PTR g_tabSubclassID = 42;
+	constexpr UINT_PTR g_comboBoxSubclassID = 42;
```
There will be 2 same values (42) for both g_tabSubclassID and g_comboBoxSubclassID.
It seems fine for the compiled PR. However, will it have some risk to have any impact?

Leonard-The-Wise replied:
No, every ID there - for any control - is 42.

donho commented:
@Leonard-The-Wise 
The compiled binary looks good.
However, the diff of PR is hard to review due to some moved blocks and reformatting.
Could you close this one, keeping the current format and create a new PR for facilitating the code review?
Thank you!

sXNA commented:
> Yeah, the button background was actually a bug, thanks for pointing. I also made it square and modified the arrow - for windows 10 of course. I'm afraid Windows 11 arrows have to stay the same for consistency though.
> 

Sorry I think you misunderstand me I mean round corners in Windows 11 and square in Windows 10.

Also I am not sure how to trigger it, but sometimes tab overflow control disappear when I close tab, it will reappear if I move mouse in place where it was.

Leonard-The-Wise commented:
@donho believe me, I didn't move blocks around. Just added some functions, renamed a few and removed some to make more sense (eg: from `tabUpDownControl` to a more generic `spinControl`, because Spin controls are not only tab exclusives - and from renderButton to renderCheckBoxOrRadioButton, because it renders checkboxes, not buttons). And then the diff became like that.

The `ComboBoxSubclass` function is **as is** and bellow `GroupboxSubclass` as in original. Means, I didn't touch it... regardless, the diff is pointing as if it moved around. Don't know how to make it better, except if i moved the new code all way to the bottom of the file, which case would make the final version ugly and disorganized. To tell the truth, the `TabUpDownSubclass` function coming after `autoSubclassAndThemeChildControls` seemed somewhat out of place already so why I moved the spinControl code to be before it.

The diff here on git seems a mess. Try reviewing it on Visual Studio. For me is a lot better there. If still not good, then I'll see what can be done then.

Thanks.

Leonard-The-Wise commented:
Ah, I also just synchonized my fork with the current version. No conflicts either.

Leonard-The-Wise commented:
> @Leonard-The-Wise
> 
> > Yeah, the button background was actually a bug, thanks for pointing. I also made it square and modified the arrow - for windows 10 of course. I'm afraid Windows 11 arrows have to stay the same for consistency though.
> 
> Sorry I think you misunderstand me I mean round corners in Windows 11 and square in Windows 10.
> 
> Also I am not sure how to trigger it, but sometimes tab overflow control disappear when I close tab, it will reappear if I move mouse in place where it was.
> 
> btw won't it be better to split this PR between buttons and tab overflow control

I understood that. Windows 10 is square, 11 is rounded.
Also, I am did no work on tab overflows here. I don't understand what bug you are talking about. For me, closing and opening tabs is always fine here.

donho commented on PowerEditor/src/NppDarkMode.cpp:
```diff
@@ -456,24 +506,30 @@ namespace NppDarkMode
 	HPEN getDarkerTextPen()               { return getTheme()._pens.darkerTextPen; }
 	HPEN getEdgePen()                     { return getTheme()._pens.edgePen; }
-	HPEN getHotEdgePen()                  { return getTheme()._pens.hotEdgePen; }
+	HPEN getHotEdgePen()				  { return getTheme()._pens.hotEdgePen; }
```
The above 2 lines (removed & added) are identical, please restore it.

donho commented on PowerEditor/src/NppDarkMode.cpp:
```diff
@@ -456,24 +506,30 @@ namespace NppDarkMode
+	HBRUSH getSofterBackgroundBrush()    { return getTheme()._brushes.softerBackground; }
+	HBRUSH getHotBackgroundBrush()       { return getTheme()._brushes.hotBackground; }
+	HBRUSH getDarkerBackgroundBrush()    { return getTheme()._brushes.pureBackground; }
+	HBRUSH getErrorBackgroundBrush()     { return getTheme()._brushes.errorBackground; }
```
The 5 above added lines are identical as the 5 removed lines, please restore them.

donho commented on PowerEditor/src/NppDarkMode.cpp:
```diff
@@ -456,24 +506,30 @@ namespace NppDarkMode
+	HBRUSH getTextBrush()				 { return getTheme()._brushes.textColorBrush; }
+	HBRUSH getDarkerTextBrush()			 { return getTheme()._brushes.darkerTextColorBrush; }
+	HBRUSH getEdgeBrush()				 { return getTheme()._brushes.edgeBrush; }
+	HBRUSH getHotEdgeBrush()			 { return getTheme()._brushes.hotEdgeBrush; }
```
The 2 above added lines are identical as the 2 removed lines, please restore them.

donho commented on PowerEditor/src/NppDarkMode.cpp:
```diff
@@ -967,65 +1303,65 @@ namespace NppDarkMode
+			break;
+		case WM_NCDESTROY:
+			RemoveWindowSubclass(hWnd, ButtonSubclass, g_buttonSubclassID);
+			delete pButtonData;
```
The 9 above added lines are identical as the 9 removed lines, please restore them.

donho reviewed (changes requested):
A lot of reformatting has been done  (no necessary change) so please redo a PR.
This PR will be closed.

donho commented:
@Leonard-The-Wise 

> Ah, I also just synchonized my fork with the current version. 

The merge is not necessary usually, but if you do it, please rebase all your commit locally then force push it. Otherwise I will have conflict to solve when I rebase the PR. By doing the suggestion above will save me the time to undo the merge and save you the time to redo a new PR.

> believe me, I didn't move blocks around. 

I do believe you. However, there are so many unnecessary modifications (reformatting by your IDE I guess) made the code review difficult. OTOH I don't think it's reasonable to include those reformatting change.
Sorry but please redo a new PR based on current master, without formatting anything.
Thank you in advance.
